feat: 부문 관리 - Admin CRUD [ISSUE-82]

jamie9504 opened the pull request:
**이슈 번호**

resolved: #82 

**작업 내용**

1. 부문 관리 - Admin CRUD
2. 부문 관리 - User R (사용중인 부문)

**테스트 작성 여부**

- [X] Test case

**주의 사항**

jamie9504 committed:
feat: SectorAcceptanceTest 작성 및 createAdmin 구현

- SectorAcceptanceTest 작성
- SectorResponse 생성
- Admin 권한이 필요하므로 UserService createAdmin 구현 및 테스트

jamie9504 committed:
feat: SectorController Mock 데이터 반환하도록 구현

- SectorController Mock 데이터 반환하도록 구현
- SectorAcceptanceTest 구현
- SpringSecurity USER일 때 가능한 기능, ADMIN일 떄도 가능하도록 변경
- User의 경우, Post를 LAZY 로딩하도록 설정

jamie9504 committed:
feat: createSector 기능 구현

- SectorRepository 구현
- SectorService에 createSector 기능 구현

jamie9504 committed:
feat: createSector, findSector 기능 구현

- SectorRepository 구현
- SectorService에 createSector, findSector 기능 구현

jamie9504 committed:
feat: findAllSector 기능 구현

- SectorService에 findAllSector 기능 구현

jamie9504 committed:
feat: updateSector 기능 구현

- SectorService에 updateSector 기능 구현
- Sector에 생성자와 수정자 추가

jamie9504 committed:
feat: deleteSector 기능 구현

- SectorService에 deleteSector 기능 구현
- SectorState Enum 추가

jamie9504 committed:
test: Sector Entity 테스트 추가

jamie9504 committed:
test: Sector Name, Description 객체 분리 및 테스트

- Sector Name, Description Embeddable 객체 분리 및 테스트
- 유효한 파라미터가 들어오지 않았을 경우 InvalidParameterException으로 처리

jamie9504 committed:
feat: Controller 실제 로직 적용

- Controller 목 데이터 반환 > 실제 로직 적용으로 변경
- /admin/이하 도메인 ADMIN ROLE일 때만 사용 가능하도록 변경
- Admin을 위한 SectorResponseForAdmin 추가
- findSector, findAllSector와 findUsedSector, findAllUsedSector 분리 - 권한별
- init-table에 sector table 추가

jamie9504 committed:
feat: 사용하지 않는 Exception 제거 및 GlobalExceptionHandler Test 추가

- InvalidParameterException를 UserInputException으로 대체
- GlobalExceptionHandler Test

jamie9504 committed:
test: Controller Test 보강

jamie9504 committed:
refactor: complict 해결

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/SectorAdminController.java:
```diff
@@ -0,0 +1,65 @@
+    @GetMapping("/{id}")
+    public ResponseEntity<SectorResponseForAdmin> findSector(@PathVariable Long id) {
+        SectorResponseForAdmin sector = sectorService.findSector(id);
+        return ResponseEntity
+            .ok(sector);
```
점이 한개라서 줄바꿈 없어도 될것같아요!

jamie9504 replied:
통일성 있게 줄바꿈으로 가겠습니다!

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/SectorAdminController.java:
```diff
@@ -0,0 +1,65 @@
+    @GetMapping
+    public ResponseEntity<List<SectorResponseForAdmin>> findAllSector() {
+        List<SectorResponseForAdmin> sectors = sectorService.findAllSector();
+        return ResponseEntity
+            .ok(sectors);
```
여기두요!

jamie9504 replied:
여기두요!

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/Sector.java:
```diff
@@ -0,0 +1,85 @@
+    private SectorDescription description;
+
+    @ManyToOne
+    @JoinColumn(name = "creator_id", nullable = false)
+    private User creator;
```
nullable 이 false 인데, 기본으로 만들어놓을 부문의 경우 creator 가 누가 될까요??
이따비 팀원이 되려나요?!

jamie9504 replied:
넹 관리자가 될거에요~!

aegis1920 replied:
기본이 관리자라면 default로 User(관리자)를 넣는 로직도 필요하지 않을까요?

jamie9504 replied:
프로그램을 통해 관리자가 추가하는 경우, 해당 서비스에서 관리자를 넣어줄 것으로 판단하여 default는 아무것도 지정하지 않았습니다.
(지정할 필요가 없다고 생각했어요)

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/Sector.java:
```diff
@@ -0,0 +1,85 @@
+    @JoinColumn(name = "creator_id", nullable = false)
+    private User creator;
+
+    @ManyToOne
+    @JoinColumn(name = "last_modifier_id", nullable = false)
+    private User lastModifier;
```
last modifier 는 부문 정보 자체를 수정한 사람인가요 마지막으로 글을 올린 사람인가요?!
부문정보 수정은 구현 안하는건줄 알고있었는데 아닌가요,,,?!

jamie9504 replied:
부문 정보 수정의 경우, 사용자가 부문을 등록했을 때는 관리자가 승인을 하니 관리자가 수정자가 되겠죠!?

:>

yelimkim98 replied:
그런 의미라면  부문이 publish 된 시각을 저장하는 걸로 가면 어떠려나요?!

jamie9504 replied:
시각은 저장될 것이지만 어떤 관리자가 승인했는지 알 필요성이 있다고 생각했고,
추후 문제가 되는 단어같은 것이 발견되면 관리자가 수정할 수도 있다고 생각했어요~!

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/Sector.java:
```diff
@@ -0,0 +1,85 @@
+    @Builder
+    public Sector(String name, String description, User creator, User lastModifier,
+        SectorState state) {
+        this.name = new SectorName(name);
+        this.description = new SectorDescription(description);
+        validate(creator, lastModifier, state);
```
validate 가 제일 위에 오는건 어떤가요?!
별 차이는 없을것 같지만 조금이라도 불필요한 일을 안할 수 있고
그게 더 보기 좋을 것 같아서요.?!

jamie9504 replied:
오 좋아요~!

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/Sector.java:
```diff
@@ -0,0 +1,85 @@
+        this.state = state;
+    }
+
+    private void validate(User creator, User lastModifier, SectorState state) {
+        Objects.requireNonNull(creator, "Creator는 Null일 수 없습니다.");
+        Objects.requireNonNull(lastModifier, "LastModifier는 Null일 수 없습니다.");
+        Objects.requireNonNull(state, "SectorState는 Null일 수 없습니다.");
+    }
```
User, SectorState 모두 가공되어 들어오는 데이터인 것 같은데 null 테스트가 없어도 되지 않을까요?!

User 생성시 Email, Password 등의 객체에 null 테스트를 구현했다가
피드백을 반영해서 뺀 적이 있었는데
그 상황과 어떤 차이가 있나요?!

jamie9504 replied:
가공되서 들어오지만(즉, 지금 로직에서는 상관없지만), 
클래스의 입장에선 생성자가 열려있기 때문에 null이 들어올 수 있다고 판단했습니다.

이따가 빙봉이랑 같이 얘기해보면 좋을 것 같네요.

jamie9504 replied:
null을 넣는 것으로 일단 얘기되었습니다!

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/SectorName.java:
```diff
@@ -0,0 +1,38 @@
+
+    SectorName(String name) {
+        validation(name);
+        this.name = name.trim()
+            .replaceAll(" +", " ")
+            .toUpperCase();
```
더하기(+)와 UpperCase 의 의미가 뭔가요?!
부문명을 어떤 식으로 입력받을 거라고 생각해서 이렇게 구현하신건지 궁금합니다!

jamie9504 replied:
1. replaceAll(" +", " ")의 경우엔 2칸 이상의 공백을 모두 한 칸의 공백으로 맞춰주는 것입니다.
2. UpperCase의 경우 영문자를 모두 대문자로 맞춰주는 것입니다.

부문명의 경우 통일감이 있어야 할 것 같아서 모두 대문자로 맞춰보았고, (unique 반영에도 용이)
공백 두 칸 이상의 경우 오타로 판단하여 모두 한 칸으로 변경시켜버렸습니다.

aegis1920 replied:
unique 반영에 용이하다는 게 성능적으로 용이하다는 건가요??

jamie9504 replied:
unique일 때 대/소문자를 구분하느냐 하지 않느냐가 DB에 따라 다른 것으로 알고있고
일부 DB에는 구분하게끔 설정하는 것이 DDL에 있었던 것으로 기억하는데,

그렇게 DB의 기능을 활용하기보다는 모두 대문자로 저장하는 것이 더 좋다고 생각했습니다.
또한, 비슷한 부문이 어떤 것은 소문자/ 어떤 것은 대문자인 것이 가독성(?)도 별로라고 생각했어요.

aegis1920 replied:
그렇다면 저도 소문자보다는 대문자가 나은 것 같네요 👍

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/SectorService.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Transactional
+    public SectorResponse createSector(SectorRequest sectorRequest) {
+        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
```
저는 유저정보 가져올 때
컨트롤러에서 Authentication 을 매개변수에 쓰고
서비스에도 이걸 전달해주는 방식으로 했었는데,
뭐가 더 좋은지는 잘 모르겠어요.
상의해서 통일하는 것도 괜찮을 것 같아요 (통일 안해도 괜찮구요!)

jamie9504 replied:
이 부분 라흐랑 얘기했었을 때,

SecurityContextHolder가 AOP 관점에서 어떤 영역에서도 접근이 되니 이렇게 사용하는 게 나을 것 같다는 이야기를 했었어요!

이렇게 사용하게 되면, Service에 Authentication을 넘기지 않아도 됩니다 :)
- 테스트가 용이해져요!

yelimkim98 replied:
아하?
그래도 내부적으로 Authentication을 사용하는데
모킹을 안해줄 수 있나요??
오히려 SecurityContextHolder부터 한줄 쭉 모킹해줘야할거같은데 아닌가요..?!

jamie9504 replied:
모킹을 하지 않고, 직접 셋팅을 할 수 있어요!

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/UserService.java:
```diff
@@ -33,14 +33,24 @@
             .nickname(new Nickname(userCreateRequest.getNickname()))
             .password(new Password(passwordEncoder.encode(userCreateRequest.getPassword())))
-            .roles(Collections.singletonList(Role.USER.getRoleName()))
+            .roles(Collections.singletonList(user.getRoleName()))
```
user 대신 userRole 같은 변수명은 어떨까요?
user 객체에서 RoleName 을 얻어오고있는 줄 알고, 어떻게한거지 싶었는데
Role 객체 이름이 user 였던거더라구요,,

클래스가 Role 클래스이긴 하지만..
```User user``` 하듯이 ```Role userRole``` 이것도 괜찮지 않을까요?!

jamie9504 replied:
변경했습니다 👍

yelimkim98 reviewed (commented):
약간의 코멘트 남깁니다!
이 다음번엔 approve 할것같네요 :)

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/Sector.java:
```diff
@@ -0,0 +1,85 @@
+
+    @ManyToOne
+    @JoinColumn(name = "last_modifier_id", nullable = false)
+    private User lastModifier;
```
저도 부문정보 수정은 없는건줄 알고 있었습니당!!

jamie9504 replied:
일반 사용자의 경우 가장 나중에 승인자가 있을테니 해당 부분 넣어놨습니다~

+ 어드민의 경우 수정을 할 수 있다고 가정했습니다.

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/SectorDescription.java:
```diff
@@ -0,0 +1,38 @@
+    private String description;
+
+    SectorDescription(String description) {
+        validation(description);
```
validate라고 동사형으로 쓰는 건 어떨까요??
지난 Post에서 검증 함수가 validateLength 등 동사형으로 되어 있어서 동사형으로 맞추면 좋을거같습니다~

jamie9504 replied:
오 좋습니다 👍

ttkmw reviewed (approved):
수고하셨습니다!
몇가지 피드백 남겼는데 참고 부탁드려요~

jamie9504 committed:
refactor: 검증 메서드 동사형으로 변경(validation > validate)

jamie9504 committed:
refactor: Role user > userRole로 변수명 변경

jamie9504 committed:
refactor: 검증 메서드 위치 변경

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/SectorAdminController.java:
```diff
@@ -0,0 +1,65 @@
+@AllArgsConstructor
+public class SectorAdminController {
+
+    private SectorService sectorService;
```
재할당 되지 않도록 final을 붙여주면 좋을 것 같아요 :)

jamie9504 replied:
오 굳!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/SectorAdminController.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<SectorResponseForAdmin> findSector(@PathVariable Long id) {
```
SectorResponseForAdmin 보다 AdminSectorResponse 가 클래스명으로 적합하지 않을까요?

jamie9504 replied:
변경했습니다~!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/SectorController.java:
```diff
@@ -0,0 +1,33 @@
+@AllArgsConstructor
+public class SectorController {
+
+    private SectorService sectorService;
```
여기두요!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/Sector.java:
```diff
@@ -0,0 +1,85 @@
+@ToString
+public class Sector extends BaseEntity {
+
+    @Embedded
+    private SectorName name;
+
+    @Embedded
+    private SectorDescription description;
```
sector 패키지 안에 있다면 필드명에 Sector를 안 붙여도 괜찮지 않을까요?

jamie9504 replied:
오호.. USER의 name과 겹쳐서 고민했는데 빼버렸습니다~!

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/config/WebSecurityConfig.java:
```diff
@@ -51,8 +51,10 @@ protected void configure(HttpSecurity http) throws Exception {
             .authorizeRequests()
             .antMatchers("/join").permitAll()
+            .antMatchers("/joinAdmin").permitAll()
             .antMatchers("/login").permitAll()
-            .antMatchers("/**").hasRole("USER")
+            .antMatchers("/admin/**").hasRole("ADMIN")
+            .antMatchers("/**").hasAnyRole("USER", "ADMIN")
```
56번 라인만 남기고 다 지워도 되지 않나요?

jamie9504 replied:
54번 라인이 없으면 Admin을 생성할 수 없어요,
57번 라인이 없으면 USER가 아니더라도 (로그인이 안되있더라도) 사용할 수 있을거에요!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/SectorState.java:
```diff
@@ -0,0 +1,17 @@
+package wooteco.team.ittabi.legenoaroundhere.domain.sector;
+
+public enum SectorState {
```
기존에 있던 State와 SectorState를 나눈 이유가 있을까요?

jamie9504 replied:
Sector에서 다뤄야하는 State가 달라서(등록요청, 승인) 구분했습니다.

aegis1920 replied:
어떤 상태가 다른지 말해줄 수 있을까요? 등록 요청이나 승인부분도 하나의 State에 추가해놓고 Sector에서만 사용해도 될 것 같아서요!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/dto/SectorResponseForAdmin.java:
```diff
@@ -0,0 +1,45 @@
+@Getter
+@EqualsAndHashCode
+@ToString
+public class SectorResponseForAdmin {
```
여기도 For을 빼고 AdminSectorResponse는 어떨까요?

jamie9504 replied:
변경했습니다~!

jamie9504 committed:
feat: Setter 검증메서드 추가

aegis1920 reviewed (commented):
제이미 고생하셨습니당 :)

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/UserController.java:
```diff
@@ -32,6 +32,15 @@ public UserController(UserService userService) {
 
+    @CrossOrigin
+    @PostMapping("/joinAdmin")
+    public ResponseEntity<Void> joinAdmin(@RequestBody UserCreateRequest userCreateRequest) {
```
joinAdmin으로 누구나 어드민 계정을 만들 수 있을 것 같아요. 어드민 프론트 페이지를 따로 만들어 관리하지 않으니 '어드민' 자체를 없애고 부문을 db로 관리하는 방법도 있을 것 같아요. 해당 부분은 다 같이 토론해보시죠!

jamie9504 replied:
이 부분 토론이 필요해요~_~!

jamie9504 committed:
refactor: 코드 리팩터링

- SectorResponseForAdmin > AdminSectorResponse
- Service 빈을 담는 변수에 final을 붙임
- SectorName, SectorDescription > Name, Description

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/SectorDescription.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private void validate(String description) {
+        if (Objects.isNull(description) || isInvalid(description.trim().replaceAll(" +", " "))) {
```
일부러 공백을 여러개 넣어서 작성할 수도 있지 않을까요? (뭔가 드립을 치려고 줄을 맞춘다던지)
글 내용을 유연하게 가져가도 좋을 것 같아요!

jamie9504 replied:
부문에 대한 한 줄 소개 같은 것으로 판단해서, 여러 공백을 막았는데....
이 부분은 추후에 도메인 관련해서(전반적인 글자 제한 등) 다시 얘기해보면 좋을 것 같아요!

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/SectorState.java:
```diff
@@ -0,0 +1,17 @@
+public enum SectorState {
+
+    PUBLISHED(true),
+    DELETED(false);
```
승인 대기인 state는 필요 없을까요?

```
승인대기 => PENDING
승인 => APPROVED
거절 => REJECTED
```
는 어떨까요?

jamie9504 replied:
승인 대기는 아직 필요 없어서... User쪽 작업할 때 추가해두면 좋을 것 같아요!

yelimkim98 replied:
우리 로직이 확실하게 펜딩상태가 필요하다는 로직이라는 생각이 들어서
미리 펜딩상태 만들어놓는것도 의미있을것같아요!

ttkmw replied:
@jamie9504 
Readme.md 파일에는 부문 상태를 ApprovalPending, Approved, Rejected로 정의했었습니다!
문서를 PUBLISHED, DELETED로 수정해주시면 좋을 것 같네요~

jamie9504 replied:
Tdd라서 일단 저렇게 두었는데, 추후 수정시 참고하겠습니다~!

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/SectorService.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    public SectorResponse findUsedSector(Long id) {
```
UsedSector 단어만 봤을 때 '사용된 부문'이라고 해석이 되는데 어떤 의미인지 와닿지가 않네요!
![image](https://user-images.githubusercontent.com/12741994/89384219-b8652c00-d738-11ea-9354-338fe2e26736.png)

등록된 부문(`enrolledSector`)은 어떨까요?

jamie9504 replied:
사용중인 부문으로 하고 싶어서.. 일단 `in use`로 변경했어요!

Hamliet commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/controller/GlobalExceptionHandlerTest.java:
```diff
@@ -0,0 +1,143 @@
+
+@SpringBootTest
+@AutoConfigureMockMvc
+class GlobalExceptionHandlerTest {
```
👍👍👍👍👍👍👍

Hamliet commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/service/SectorServiceTest.java:
```diff
@@ -0,0 +1,241 @@
+
+    @DisplayName("Sector 생성 - 성공")
+    @Test
+    void createSector_NotDuplicateName_Success() {
```
성공하는 케이스에 대해 중복되지 않았다고 명시하는것보다 없는것이 나을 것 같아요!
`createSector_NotDuplicatedName_Success` -> `createSector_Success`
validate 통과한 항목이 Duplicate뿐만이 아니니까요! (SectorName - 길이, 공백...  SectorDescription - 길이, 공백...)

jamie9504 replied:
오 그렇네용!! 수정했습니다!

Hamliet reviewed (commented):
안녕하세요 제이미! 깔끔하게 잘 구현해주셨네요!!
몇 가지 코멘트 남겼으니 확인 부탁드려요~!!

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/sector/Description.java:
```diff
@@ -0,0 +1,38 @@
+
+    private void validate(String description) {
+        if (Objects.isNull(description) || isInvalid(description.trim().replaceAll(" +", " "))) {
+            throw new UserInputException("SectorDescription이 유효하지 않습니다.");
```
예전에 리뷰하려다가 놓친 부분인데요,
`UserInputException`이 현상에 대한 의미가 담겨져있지 않은 것 같아요.
`WrongInputException`이나 `WrongUserInputException` 등이 좋을 것 같아요.

jamie9504 replied:
좋습니다~!

jamie9504 committed:
refactor: Exception명 변경(UserInputException > InputException)

jamie9504 committed:
refactor: 메서드명 변경

jamie9504 committed:
refactor: 사용하지 않는 접근 권한 제한자 제거

Hamliet reviewed (approved):
수고하셨습니다 제이미! 이만 Approve할게요!! :)

aegis1920 reviewed (approved):
고생했습니다 제이미 :)
이만 Approve 할게요!!

yelimkim98 reviewed (approved):
달아논 줄 알았던 코멘트들이 펜딩상태로 남아있었네요..!

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#82

# Conflicts:
#	back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/config/WebSecurityConfig.java
#	back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/UserController.java

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#82

# Conflicts:
#	back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/Image.java
#	back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/service/UserServiceTest.java
#	back-end/legeno-around-here/src/test/resources/init-table.sql

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#82

# Conflicts:
#	back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/Image.java
#	back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/service/UserServiceTest.java
#	back-end/legeno-around-here/src/test/resources/init-table.sql
